overlord/hookstate/ctlcmd/model.go

```diff
@@ -0,0 +1,280 @@
+	"github.com/snapcore/snapd/snap"
+)
+
+var (
+	shortModelHelp = i18n.G("Get the active model for this device")
+	longModelHelp  = i18n.G(`
+The model command returns the active model assertion information for this
+device.
+
+By default, the model identification information is presented in a structured,
+yaml-like format, but this can be changed to json by using the --json flag.
+`)
+)
+
+func init() {
+	addCommand("model", shortModelHelp, longModelHelp, func() command { return &modelCommand{} })
+}
+
+type modelCommand struct {
+	baseCommand
+	Assertion bool `long:"assertion"`
+	Json      bool `long:"json"`
+}
+
+type modelCommandFormatter struct {
+	snapInfo *snap.Info
+}
+
+// GetEscapedDash implements part of the clientutil.ModelFormatter interface
+func (mf modelCommandFormatter) GetEscapedDash() string {
+	return "--"
+}
+
+// LongPublisher implements part of the clientutil.ModelFormatter interface
+// essentially this functions reimplements the same logic as cmd/snap/color.go
+// but without all the fancy formatting as we don't need it. We also will not use
+// the unicode characters as the output is formatted for API rather than human
+// consumption
+func (mf modelCommandFormatter) LongPublisher(storeAccountID string) string {
+	if mf.snapInfo == nil || mf.snapInfo.Publisher.DisplayName == "" {
+		return mf.GetEscapedDash()
+	}
+
+	storeAccount := mf.snapInfo.Publisher
+	var badge string
+	switch storeAccount.Validation {
+	case "verified":
+		badge = "**"
+	case "starred":
+		badge = "*"
+	}
+
+	// NOTE this makes e.g. 'Potato' == 'potato', and 'Potato Team' == 'potato-team',
+	// but 'Potato Team' != 'potatoteam', 'Potato Inc.' != 'potato' (in fact 'Potato Inc.' != 'potato-inc')
+	if strings.EqualFold(strings.Replace(storeAccount.Username, "-", " ", -1), storeAccount.DisplayName) {
+		return storeAccount.DisplayName + badge
+	}
+	return fmt.Sprintf("%s (%s%s)", storeAccount.DisplayName, storeAccount.Username, badge)
+}
+
+func (c *modelCommand) newTabWriter(output io.Writer) *tabwriter.Writer {
+	minWidth := 2
+	tabWidth := 2
+	padding := 2
+	padchar := byte(' ')
+	return tabwriter.NewWriter(output, minWidth, tabWidth, padding, padchar, 0)
+}
+
+// reportError prints the error message to stderr
+func (c *modelCommand) reportError(format string, a ...interface{}) {
+	w := c.newTabWriter(c.stderr)
+	fmt.Fprintf(w, format, a...)
+	w.Flush()
+}
+
+// hasSnapdControl returns true if the requesting snap has the snapd-control plug
+// and only if it is connected as well.
+var hasSnapdControlInterface = func(st *state.State, snapName string) (bool, error) {
+	conns, err := ifacestate.ConnectionStates(st)
+	if err != nil {
+		return false, err
+	}
+	for refStr, connState := range conns {
+		if connState.Undesired || connState.Interface != "snapd-control" {
+			continue
+		}
+		connRef, err := interfaces.ParseConnRef(refStr)
+		if err != nil {
+			return false, err
+		}
+		if connRef.PlugRef.Snap == snapName {
+			return true, nil
+		}
```

Comment:
this loop is not exercised, refresh_test.go might have idea how to setup things to test it

Reply:
Good point about refresh_test - it turned out to be pretty straight forward and also removed the need for me to mock it.